proxy/src/error.rs

```diff
@@ -0,0 +1,155 @@
+use juniper::{
+    FieldError, IntoFieldError, Value,
+};
+
+use radicle_registry_client;
+use radicle_registry_client::{Error as RegistryError};
+use radicle_surf::{
+    file_system::{Path},
+};
+
+use crate::source::{Error as SourceError};
+
+#[derive(Debug)]
+/// Enumerable of expected error types.
+pub enum Error {
+    /// File at a given path was irretrievable.
+    FileNotFound(Path),
+    /// Directory at a given path was irretrievable.
+    DirectoryNotFound(Path),
+    /// Project name exceeded 32 characters.
+    BadProjectName(String),
+    /// Project domain exceeded 32 characters.
+    BadProjectDomain(String),
+    /// Errors originating in radicle-surf's Git adapter.
+    Git(radicle_surf::git::GitError),
+    /// Registry client errors.
+    Registry(RegistryError),
+    /// VCS Browser could not find the last commit of a branch.
+    LastCommitNotFound(Path),
+}
+
+impl From<radicle_surf::git::GitError> for Error {
+    fn from(git_error: radicle_surf::git::GitError) -> Self {
+        Self::Git(git_error)
+    }
+}
+
+impl From<SourceError> for Error {
+    fn from(source_error: SourceError) -> Self {
+        match source_error {
+            SourceError::BadProjectName(name) => Self::BadProjectName(name),
+            SourceError::BadProjectDomain(domain) => Self::BadProjectDomain(domain),
+            SourceError::Registry(error) => Self::Registry(error),
+        }
+    }
+}
+
+impl IntoFieldError for Error {
+    fn into_field_error(self) -> FieldError {
+        match self {
+            Self::Git(git_error) => {
+                match &git_error {
```

Comment:
These sub-matches are extending the function body quite a bit, maybe we can split them out. Or if possible implement `IntoFieldError` for the contained errors.

Reply:
I can't implement IntoFieldError directly on the Git error types because they belong to radicle_surf, not the proxy. Instead I broke the matching out into a `convert_git_error_to_field_error` function -- same for RegistryError, which belongs to radicle_registry_client.